pkg/frontend/querymiddleware/split_and_cache.go

```diff
@@ -365,23 +375,59 @@ func (s *splitAndCacheMiddleware) fetchCacheExtents(ctx context.Context, keys []
 			continue
 		}
 
-		extents[keyIdx] = resp.Extents
+		extents[keyIdx] = make([]Extent, 0, len(resp.Extents))
+
+		// Filter out extents that are outside TTL.
+		for ix := range resp.Extents {
+			if resp.Extents[ix].QueryTime == 0 {
+				// If we don't know the query time, it's too old.
+				extentsOutOfTTL++
+				continue
```

Comment:
I haven't considered it. It would mean that "per-tenant TTL" feature only starts working correctly 7 days after updating your Mimir installation. On the other hand, people can always flush cache manually, if needed, so I'm fine with that approach too.

Reply:
I've implemented this, but the previous logic was easier to understand, especially when merging multiple extents.